### PR TITLE
Fixed some `Transform3DGizmo` behavior to make more consistent

### LINKED ISSUE #18987: Setting a global rotate and scale may lead to a broken Basis
Spatial has global rotate and scale methods, but those may not work in general when a node in the chain (apart from the node at the very end of the chain) contains a non-uniform scale.

Consider a chain two just two nodes, and just consider the basis part for simplicity; the global basis is Bg = (Rp.Sp)(R.S) where Rp and Sp are scale and rotation of the parent node. When you apply a global scale S', the new global basis becomes Bg -> S' Bg = S'.(Rp.Sp)(R.S), and the current code, `set_global_transform`, updates the node's local scale as AffineInverse[(Rp.Sp)].S'.(Rp.Sp).(R.S). When S' and Sp are uniform scales, they're simply proportional to identity and can be moved around freely like AffineInverse[(Rp.Sp)].S'.(Rp.Sp).(R.S) = AffineInverse[(Rp.Sp)].(Rp.Sp).(R.S'.S) resulting in (R.S'.S).

But when Sp or S' is a nonuniform scale, the resulting basis is no longer a rotation-scale matrix in the form R.S (but rather M.R.S where M is S' in the inverted parent-frame), which is an assumption made almost everywhere in the code base ---it can rather for example involve shearing, in which case  `set_global_transform`  can break stuff *silently* when used.

I don't see a way this can be fixed, but we can add checks (wrapped by MATH_CHECKS) and warn the user when this happens.

### PR DESCRIPTION
**Pull request by TokageItLab**
~**Note: According to the discussion with @reduz on the contributors chat, it was decided to go ahead with the implementation of the 8 rotation modes (euler combinations, quaternion and custom basis). After that is merged, we will consider this change again.**~ #54084 has been merged.

---
This PR makes some changes in Gizmo.

1. Implement non-orthogonal Gizmo
2. Fixes for moving and scaling when multiple objects are selected in local transform mode
3. ~~Implement a mode in global transform mode that keeps orthogonality (and add an orthogonal option to Node3D for this purpose)~~ Superseded by #54084.

It addresses the problems those were mentioned in https://github.com/godotengine/godot/pull/51164#issuecomment-891551190. Fixed #18987.

---
### Implement non-orthogonal Gizmo

Currently, even when an object's axis is non-orthogonal in the 3D view, Transform3DGizmo tries to keep it orthogonal, so local transformations along the axis are not performed correctly in such cases. This will fix that.

スクリーンショット 2021-10-12 5 05 48

---
### Fixes for moving and scaling when multiple objects are selected in local transform mode

In Local Transform mode, when multiple objects are selected, rotation correctly takes into account the local axis, but translation and scaling do not. This will make these a consistent system.

https://user-images.githubusercontent.com/61938263/136849732-f6916501-b561-4a35-90b8-f61c453091b7.mov

---
### Implement a mode in global transform mode that keeps orthogonality (and add an orthogonal option to Node3D for this purpose)

The current global transformation causes shear in basis. Shear is not allowed in glTF and is not compatible with the compression track that @reduz is working on, especially in bone animation. So this will implement global scaling that does not cause shear like in Blender.

https://user-images.githubusercontent.com/61938263/136850064-8cd89c74-ee49-4786-b9b4-23fc204309be.mov

### CODE REVIEW

**Comment by TokageItLab** [comment]
It may be necessary to orthogonalize with an importer or migrator for objects that are currently in shear regardless of the rotation mode is not basis.

File: editor/plugins/node_3d_editor_plugin.cpp [hunk 1]
```diff
@@ -1876,14 +1881,14 @@ void Node3DEditorViewport::_sinput(const Ref<InputEvent> &p_event) {
 							if (se->gizmo.is_valid()) {
 								for (KeyValue<int, Transform3D> &GE : se->subgizmos) {
 									Transform3D xform = GE.value;
-									Transform3D new_xform = _compute_transform(TRANSFORM_SCALE, se->original * xform, xform, motion, snap, local_coords);
+									Transform3D new_xform = _compute_transform(TRANSFORM_SCALE, se->original * xform, xform, motion, snap, local_coords, true); // Force orthogonal with subgizmo.
```
Comment (TokageItLab) [inline editor/plugins/node_3d_editor_plugin.cpp:1884]: Currently SubGizmo is forcing orthogonal, but I'm wondering how to make this work better.
At least it's independent of Node3D's RotationMode, but if you want to manage it on a per-SubGizmo make orthogonal, we need to make the SubGizmo store more info than just Transform.

**Comment by TokageItLab** [comment]
When multiple objects are deformed, there is a possibility that the Gizmo will move to an unexpected position due to a shift in the center.

It's mathematically and programmatically consistent, but it's may hard to understand, so it might be better to visualize the center using a bounding box or a dotted line from each origin to the center.

Or, in such a case, not moving the Gizmo during the transformation may be a solution.

**Comment by TokageItLab** [comment]
Added `scale_orthogonal()` in Basis, it is also necessary for retargeting nodes, since shearing is not allowed in bones.

@JFonS Please review when you have time.

**Review by JFonS — APPROVED** [review]
Looks great :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**